File: ruby/generate/generate.rb
# Copyright (c) 2019 Yodlee, Inc. All Rights Reserved.
# Licensed under the MIT License. See `LICENSE` for details.

# USAGE
#
#   generate.rb -k path/to/private.key -i issuer-id [-u username]
#

require 'jwt'
require 'optparse'

usage = "Usage: generate.rb -k path/to/private.key -i issuer-id [-u username]"

options = {}
OptionParser.new do |opts|
  opts.on("-h", "--help", "show help") do |h|
    puts usage
    exit
  end

  opts.on("-k KEY", "--key KEY", "Path to private key") do |key|
    options[:key] = key
  end

  opts.on("-i","--issuer-id issuer-id","Your Issuer Id") do |issuer_id|
    options[:issuer_id] = issuer_id
  end
  
  opts.on("-u","--username username", "A user name") do |username|
    options[:username] = username
  end
end.parse!

errors = []
[:key, :issuer_id].each do |key|
  unless options.has_key?(key)
    errors << "Option parameter %s is required" % key
  end
end

if errors.length > 0
  puts errors
  puts usage
  exit
end

begin
  rsa_private_key = OpenSSL::PKey::RSA.new File.read options[:key]
rescue Errno::ENOENT => e
  puts "Caught the exception: #{e}"
  puts usage
  exit -1
end

current_time = Time.now.to_i

payload = {}
payload['iss'] = options[:issuer_id]
payload['iat'] = current_time
payload['exp'] = current_time + 1800

if options.key?(:username)
  payload['sub'] = options[:username]
end

token = JWT.encode payload, rsa_private_key, 'RS512', { typ: 'JWT' }

puts token
# puts JWT.decode token, rsa_public_key, true, { :algorithm => 'RS512' }

File: python/generate/generate.py
# Copyright (c) 2019 Yodlee, Inc. All Rights Reserved.
# Licensed under the MIT License. See `LICENSE` for details.

"""Usage:  generate.py [--key=<str>] [--issuer-id=<str>] [--username=<str>]
           generate.py (-h | --help)

Options:
  --key=<str>       path to private key
  --issuer-id=<str>    issuer-id id from api connsole
  --username=<str>  username if generating user token
  -h --help         Show this screen.
"""

import jwt
import time
from docopt import docopt

def generate(arguments):
    del arguments['--help']

    proceed = True
    for k, v in arguments.items():
        if k == '--username':
            continue

        if v is None:
            print(f"parameter {k} is required")
            proceed = False

    if proceed is False:
        return 'sorry'

    privateKeyPath = arguments['--key']

    with open(privateKeyPath) as f:
        privateKey = f.read()

    currentTime = int(time.time())

    payload = {}
    payload['iat'] = currentTime
    payload['exp'] = currentTime + 1800
    payload['iss'] = arguments['--issuer-id']

    if '--username' in arguments and arguments['--username'] is not None:
        payload['sub'] = arguments['--username']

    # RS512 requires $ pip install cryptography
    encoded_jwt = jwt.encode(payload, privateKey, algorithm='RS512')
    return encoded_jwt

if __name__ == '__main__':
    arguments = docopt(__doc__)

    print(generate(arguments).decode("utf-8"))
